platforms/core-configuration/dependency-management-serialization-codecs/src/main/kotlin/org/gradle/internal/serialize/codecs/dm/ImmutableAttributesSchemaCodec.kt

```diff
@@ -0,0 +1,198 @@
+
+class ImmutableAttributesSchemaCodec(
+    private val instantiatorFactory: InstantiatorFactory,
+    private val schemaFactory: ImmutableAttributesSchemaFactory
+) : Codec<ImmutableAttributesSchema> {
+
+    override suspend fun WriteContext.encode(value: ImmutableAttributesSchema) {
+        // Attributes schemas are interned. See `ImmutableAttributesSchemaFactory`
+        encodePreservingSharedIdentityOf(value) {
+            writeSchema(it)
+        }
+    }
+
+    override suspend fun ReadContext.decode(): ImmutableAttributesSchema {
+        return decodePreservingSharedIdentity {
+            readSchema(instantiatorFactory.inject(), schemaFactory)
+        }
+    }
+}
+
+
+const val INSTANTIATING_ACTION_TYPE = 0
+const val OPAQUE_ACTION_TYPE = 1
+
+
+private
+suspend fun WriteContext.writeSchema(schema: ImmutableAttributesSchema) {
+    writeCollection(schema.attributes) { attribute ->
+        writeAttribute(attribute)
+        val value = schema.getStrategy(attribute)!!
+        writeStrategy(value)
+    }
+    writeCollection(schema.attributeDisambiguationPrecedence) {
+        writeAttribute(it)
+    }
+}
+
+private
+suspend fun <T: Any> WriteContext.writeStrategy(value: ImmutableAttributesSchema.ImmutableAttributeMatchingStrategy<T>) {
+    writeCollection(value.compatibilityRules) {
+        when (it) {
+            is InstantiatingAction -> {
+                writeSmallInt(INSTANTIATING_ACTION_TYPE)
+                writeInstantiatingAction(it)
+            }
+            else -> {
+                writeSmallInt(OPAQUE_ACTION_TYPE)
+                write(it)
+            }
+        }
+    }
+    writeCollection(value.disambiguationRules) {
+        when (it) {
+            is InstantiatingAction -> {
+                writeSmallInt(INSTANTIATING_ACTION_TYPE)
+                writeInstantiatingAction(it)
+            }
+            else -> {
+                writeSmallInt(OPAQUE_ACTION_TYPE)
+                write(it)
+            }
+        }
+    }
+}
+
+
+private
+suspend fun WriteContext.writeInstantiatingAction(action: InstantiatingAction<*>) {
+    val rules = action.rules.configurableRules
+    require(rules.size == 1)
+    val rule = rules[0]
+
+    writeClass(rule.ruleClass)
+    write(rule.ruleParams)
+}
+
+
+private
+fun WriteContext.writeAttribute(attribute: Attribute<*>) {
+    writeString(attribute.name)
+    writeClass(attribute.type)
+}
+
+
+private
+suspend fun ReadContext.readSchema(instantiator: Instantiator, schemaFactory: ImmutableAttributesSchemaFactory): ImmutableAttributesSchema {
+    val strategies = readList {
+        val attribute = readAttribute()
+        readStrategy(attribute, instantiator)
+    }.associate { it }
+
+    val precedence = readList {
+        readAttribute()
+    }
+
+    return schemaFactory.create(
+        ImmutableMap.copyOf(strategies),
+        ImmutableList.copyOf(precedence)
+    )
+}
+
+
+private suspend fun <T : Any> ReadContext.readStrategy(
```

Comment:
💅 Consistent format for visibility modifier

Reply:
Put the `private` on a new line. 

IMO if this format is something we're going to enforce, it should be enforced programatically with `sanityCheck`.

I personally don't see the benefit of putting these on a new line